File: web/assembly/diag.ts
// Structured failure, for every layer.
//
// PLAN.md §4.4: one shape for parse errors, type conflicts, limits and corrupt
// messages alike. Nothing here throws or aborts — an AssemblyScript trap reaches
// the host as an RuntimeError carrying no path and no offset, which is exactly
// what a caller cannot act on.

import { Writer } from './bytes'

export const D_NONE: i32 = 0
export const D_SYNTAX: i32 = 1
export const D_NUMBER: i32 = 2
export const D_LIMIT: i32 = 3
export const D_CONFLICT: i32 = 4
export const D_CORRUPT: i32 = 5
export const D_UNSUPPORTED: i32 = 6

export class Diag {
  code: i32 = D_NONE
  offset: i32 = -1
  line: i32 = -1
  path: string = ''
  message: string = ''
  warnings: Array<string> = []

  @inline get ok(): bool {
    return this.code == D_NONE
  }

  /** First failure wins: later ones are usually consequences of the first. */
  fail(code: i32, offset: i32, path: string, message: string): void {
    if (this.code != D_NONE) return
    this.code = code
    this.offset = offset
    this.path = path
    this.message = message
  }

  warn(message: string): void {
    this.warnings.push(message)
  }

  reset(): void {
    this.code = D_NONE
    this.offset = -1
    this.line = -1
    this.path = ''
    this.message = ''
    this.warnings = []
  }

  /** The diagnostic as JSON: code, offset, line, path, message, warnings. */
  encode(): Uint8Array {
    const out = new Writer(256)
    writeText(out, '{"code":')
    writeText(out, this.code.toString())
    writeText(out, ',"offset":')
    writeText(out, this.offset.toString())
    writeText(out, ',"line":')
    writeText(out, this.line.toString())
    writeText(out, ',"path":')
    writeJSONText(out, this.path)
    writeText(out, ',"message":')
    writeJSONText(out, this.message)
    writeText(out, ',"warnings":[')
    for (let index = 0; index < this.warnings.length; index++) {
      if (index > 0) out.writeByte(0x2c)
      writeJSONText(out, unchecked(this.warnings[index]))
    }
    writeText(out, ']}')
    return out.take()
  }
}

function writeText(out: Writer, text: string): void {
  const bytes = Uint8Array.wrap(String.UTF8.encode(text, false))
  out.writeBytes(bytes, 0, bytes.length)
}

/** A JSON string literal. Diagnostics are the only strings this writes, so the
 * escape set is the minimum that keeps the envelope parseable rather than
 * encoding/json's — jsontext.ts is where that one lives. */
function writeJSONText(out: Writer, text: string): void {
  const bytes = Uint8Array.wrap(String.UTF8.encode(text, false))
  out.writeByte(0x22)
  for (let index = 0; index < bytes.length; index++) {
    const character = unchecked(bytes[index])
    if (character == 0x22 || character == 0x5c) {
      out.writeByte(0x5c)
      out.writeByte(character)
    } else if (character == 0x0a) {
      out.writeByte(0x5c)
      out.writeByte(0x6e)
    } else if (character < 0x20) {
      writeText(out, '\\u00')
      out.writeByte(hexDigit(character >> 4))
      out.writeByte(hexDigit(character & 0x0f))
    } else {
      out.writeByte(character)
    }
  }
  out.writeByte(0x22)
}

@inline
function hexDigit(value: u8): u8 {
  return value < 10 ? 0x30 + value : 0x61 + (value - 10)
}

/**
 * 1-based line of a byte offset. Computed only when reporting, so the parser
 * does not carry a line counter through its hot loop.
 */
export function lineOf(src: Uint8Array, offset: i32): i32 {
  let line = 1
  const end = offset < src.length ? offset : src.length
  for (let i = 0; i < end; i++) {
    if (unchecked(src[i]) == 0x0a) line++
  }
  return line
}

File: rust/src/diag.rs
//! Structured failure, for the encode side.
//!
//! Mirrors `web/assembly/diag.ts`. One shape for parse errors, type conflicts,
//! limits and corrupt messages alike, carrying a byte offset and a path into the
//! document so a caller can point at what was wrong rather than only say that
//! something was.
//!
//! # Why not [`crate::Error`]
//!
//! `Error` is the decoder's, and the note at the top of `error.rs` says why
//! every variant of it is a read-side failure: a writer trusts its own program.
//! That stops being true the moment the thing being encoded is a document
//! somebody typed. `{"a":1,"a":"two"}` is not a bug in this crate and not a
//! corrupt message — it is a type conflict at byte 9 — and an enum with no room
//! for the 9 could only report it as an enum variant.
//!
//! Nothing here allocates until something has already failed.

use alloc::string::{String, ToString};
use alloc::vec::Vec;

pub const D_NONE: i32 = 0;
/// The document is not JSON.
pub const D_SYNTAX: i32 = 1;
/// A number the format cannot carry exactly.
pub const D_NUMBER: i32 = 2;
/// A document past one of the encoder's bounds.
pub const D_LIMIT: i32 = 3;
/// One key holding two types, or an array the format has no shape for.
pub const D_CONFLICT: i32 = 4;
/// A message that does not read back as what went in.
pub const D_CORRUPT: i32 = 5;
/// A shape the encoder declines rather than encodes badly.
pub const D_UNSUPPORTED: i32 = 6;

/// What failed, where, and on the way to which field.
#[derive(Debug, Clone, Default)]
pub struct Diag {
    pub code: i32,
    /// Byte offset into the input, or `-1` when the failure has no position.
    pub offset: i32,
    /// 1-based line, filled in by [`Diag::locate`] when reporting.
    pub line: i32,
    /// Dotted path to the value, for the failures that have one.
    pub path: String,
    pub message: String,
    pub warnings: Vec<String>,
}

impl Diag {
    #[must_use]
    pub fn new() -> Self {
        Self {
            code: D_NONE,
            offset: -1,
            line: -1,
            ..Self::default()
        }
    }

    #[must_use]
    pub fn ok(&self) -> bool {
        self.code == D_NONE
    }

    /// First failure wins: later ones are usually consequences of the first, and
    /// the caller wants the cause rather than the last symptom.
    pub fn fail(&mut self, code: i32, offset: i32, path: &str, message: impl Into<String>) {
        if self.code != D_NONE {
            return;
        }
        self.code = code;
        self.offset = offset;
        self.path = path.to_string();
        self.message = message.into();
    }

    pub fn warn(&mut self, message: impl Into<String>) {
        self.warnings.push(message.into());
    }

    pub fn reset(&mut self) {
        *self = Self::new();
    }

    /// Fills in [`Diag::line`] from the source the offset points into. Done here
    /// rather than at the failure, so the scanner carries no line counter
    /// through its hot loop.
    pub fn locate(&mut self, src: &[u8]) {
        self.line = if self.offset < 0 {
            -1
        } else {
            line_of(src, self.offset as usize)
        };
    }

    /// The diagnostic as JSON: code, offset, line, path, message, warnings.
    ///
    /// The same envelope `web/assembly/diag.ts` writes and `rust/wasm` writes
    /// for a decode failure, so one host-side reader serves every case.
    #[must_use]
    pub fn encode(&self) -> Vec<u8> {
        let mut out = Vec::with_capacity(128 + self.message.len());
        out.extend_from_slice(br#"{"code":"#);
        crate::json::write_i64(&mut out, i64::from(self.code));
        out.extend_from_slice(br#","offset":"#);
        crate::json::write_i64(&mut out, i64::from(self.offset));
        out.extend_from_slice(br#","line":"#);
        crate::json::write_i64(&mut out, i64::from(self.line));
        out.extend_from_slice(br#","path":"#);
        crate::json::write_json_str(&mut out, &self.path);
        out.extend_from_slice(br#","message":"#);
        crate::json::write_json_str(&mut out, &self.message);
        out.extend_from_slice(br#","warnings":["#);
        for (at, warning) in self.warnings.iter().enumerate() {
            if at > 0 {
                out.push(b',');
            }
            crate::json::write_json_str(&mut out, warning);
        }
        out.extend_from_slice(b"]}");
        out
    }
}

/// The 1-based line a byte offset falls on.
#[must_use]
pub fn line_of(src: &[u8], offset: usize) -> i32 {
    let end = offset.min(src.len());
    let mut line = 1i32;
    for &byte in &src[..end] {
        if byte == b'\n' {
            line += 1;
        }
    }
    line
}
